letsencrypt-auto-source/letsencrypt-auto

```diff
@@ -488,13 +475,71 @@ BootstrapRpmCommon() {
   BootstrapRpmCommonBase "$python_pkgs"
 }
 
+# If new packages are installed by BootstrapRpmPython3 below, this version
+# number must be increased.
+BOOTSTRAP_RPM_PYTHON3_LEGACY_VERSION=1
+
+EnablePython36SCL() {
+  if "$EXISTS" python3.6 > /dev/null 2> /dev/null; then
+      return 0
+  fi
+  if ! scl --list 2>/dev/null | grep -q rh-python36; then
+      return 0
+  fi
+  set +e
+  . scl_source enable rh-python36
+  set -e
+}
+
+BootstrapRpmPython3Legacy() {
+  # Tested with:
+  #   - CentOS 6
+
+  InitializeRPMCommonBase
+
+  if ! "${TOOL}" list rh-python36 >/dev/null 2>&1; then
+    echo "To use Certbot on this operating system, packages from the SCL repository need to be installed."
+    if ! "${TOOL}" list centos-release-scl >/dev/null 2>&1; then
+      error "Enable the SCL repository and try running Certbot again."
+      exit 1
```

Comment:
The process breaks here for RHEL 6 (even though `$ASSUME_YES` is `1`) and because of this, it would break the renewal for all RHEL6 and RHEL7 systems without SCL repositories pre-enabled.

For RHEL 6 the SCL repository is enabled by:
`subscription-manager repos --enable=rhel-server-rhscl-6-rpms`

Reply:
Are you sure for RHEL 7? Because this bootstrap function is supposed to be called only when python 2.7 is not available: rhel 7 has python 2.7.

I start to implement appropriate logic for rhel 6.

Reply:
Oh, you are right. This is not the case for RHEL 7 then. I didn't test with it.

Reply:
I think most users wouldn't be affected by this due to the code [here](https://github.com/certbot/certbot/blob/c2480b29f7e289a0d3db3d8e7639b909b9069ace/letsencrypt-auto-source/letsencrypt-auto#L1082), however, breaking updates ever in this way isn't great. Thanks for flagging this @joohoi.

We can maybe fix the problem automatically on RHEL 6 (although there were some issues in the past described in https://github.com/certbot/website/pull/286), but I'm not sure we can do this for all RHEL variants. At least when quickly testing in Docker, neither Oracle Linux 6 or Scientific Linux 6 has neither `subscription-manager` or a `centos-release-scl` package. We can maybe find a `yum-config-manager` command that works, but I couldn't quickly do so.

If we add any branches of code that isn't run on CentOS 6, I think we should probably set up automated tests for the code which we don't currently have. I'll also share that my numbers from 5 months ago about the number of users we have on these OSes which is less than 1000 RHEL 6 users, less than 500 SL 6 users, and less than 250 OL 6 users.

If installing SCL on all of these systems and setting up testing isn't feasible, I think our next best option is to add the logic to check in advance if we think we can noninteractively rebootstrap and if we cannot, refuse to upgrade Certbot with a warning until certbot-auto is run attached to a terminal or SCL is enabled. I have some ideas about how to accomplish this if we go this route and it isn't pretty, but I think this is my preferred approach.

Unless someone feels strongly that we should just let certbot-auto break for a percentage of the 1750 users, I think this PR is going to miss the release tomorrow which I am personally totally OK with because it means we don't have to rush changes here.

Please let me know what you all think!

Reply:
Anyway yes, I think it is better to not try rushing this PR into the release. Nothing is broken for now and we covered new RedHat systems.

About the solution you propose to refuse upgrade if non-interactive and SCL is not enable, I am not convinced. Maybe with a concrete idea it would be better, but for what I see so far:
* either this is a logic run specifically on RHEL 6, and for the same reasons you explained, dedicated tests should be added
* either this is a global logic, testable from another OS, but this is not that great because we are not testing the real case, and the increase of regressions as side-effects is increased for other systems.

You are saying that RHEL 6, SL 6 and OL 6 are a very small portion of our users. I think the support we propose here must be in balance with it. Setting specific farm tests (either for your preferred option, or by explicit SCL enabling logic for each OS) is complex and time consuming. I do not think it is worth it.

That said, without explicit dedicated tests, we should choose the most pratical solutions for these users. And in this case, proposing auto-bootstrapping for RHEL+SL+OL is better than blocking the upgrade until a manual action is done. It does not seems too much in that extend to put 3 dedicated branches in `rpm_python3_legacy.sh` to enable SCL for these three distributions, and ensure that it is executed only for them (testing the availability of dedicated binary, or dedicated package name, seems quite strong to make the difference between OSes).

Reply:
That's a good point that testing the logic I describe with my alternative approach would likely require setting up tests on a new system too. I think the difference is we'd just need to set up tests on one of RHEL 6, SL 6 or OL 6 rather than multiple of them (plus any RHEL 6 variants we're forgetting about).

If we can be confident that we can automatically enable SCL on all RHEL 6 based systems reliably without a lot of work, I think that's great. I just worry that it's pretty hard to write and test this due to all the variables here such as [RHEL 6 historically not having subscription-manager](https://github.com/certbot/website/pull/286), [other packages/repositories having to be installed/enabled](https://docs.oracle.com/en/operating-systems/oracle-linux/scl-user/section_e3v_nbl_cr.html), or how to enable SCL being distro specific such as `centos-release-scl`.

I worry that if there's not a catch-all way to do this, some users are going to slip through the cracks and if that's the case, maybe it's better for us to focus on improving this behavior rather than writing multiple pieces of potentially complicated custom logic for groups of less than 1000 users.

Reply:
Parsing the links you provided, and going further, I understand now what you meant about the fragmentation of SCL configuration on RHEL-flavor OSes. Yes, I do not see a clear way to have a consistent bootstrap, and it may even break systems.

What was your idea to avoid certbot unattended upgrades in these conditions ?

Reply:
I was thinking we would effectively tweak the logic at https://github.com/certbot/certbot/blob/f1e580dcaf16391f06a5420540b69d00c3c67ba0/letsencrypt-auto-source/letsencrypt-auto#L1133-L1158

I've had a few ideas on how to do this including duplicating a similar version of this logic in the [code that determines how to bootstrap for RHEL 6](https://github.com/certbot/certbot/blob/f1e580dcaf16391f06a5420540b69d00c3c67ba0/letsencrypt-auto-source/letsencrypt-auto#L812), but the idea I personally like best so far is to create yet another global environment variable called something like `INTERACTIVE_BOOTSTRAP`.

The way this would work is when setting the other bootstrap variables for RHEL 6 based systems in my previous link, we also call a function that checks if either a `rh-python36` or `centos-release-scl` package is available with `yum`. If either is, this variable is set to `0` or is unset but if both are unavailable, this variable is set to `1`.

Then, the logic in my first link in this comment would look something like:

```
if [ "$NONINTERACTIVE" = 1 -a "$INTERACTIVE_BOOTSTRAP" != 1 ]; then
  CAN_REBOOTSTRAP=1
fi
if [ \( "$CAN_REBOOTSTRAP" = 1 \) -o \( \( -t 0 \) -a \( -t 1 \) \) ]; then
  if [ -d "$VENV_PATH" ]; then
    rm -rf "$VENV_PATH"
  fi
  ...
```
We'd also need to check `INTERACTIVE_BOOTSTRAP` when generating the error message about needing to rebootstrap to avoid telling people they can just use `--non-interactive` which won't work.

To test this, I think it might be easiest to set up letsencrypt-auto tests in Docker on [SL6](https://hub.docker.com/r/scientificlinux/sl/) or [OL6](https://hub.docker.com/_/oraclelinux/). I looked into setting up RHEL 6 test farm tests before and it required either some custom code or deviating from the standard AMIs to get EPEL and SCL, but in theory that'd work too.

Reply:
I added this logic about conditional rebootstrap. I will add dedicated tests tomorrow.